torch/autograd/graph.py

```diff
@@ -113,16 +113,16 @@ class save_on_cpu():
 
     """
     def __init__(self, pin_memory=False):
+        self.cuda_avail = torch.cuda.is_available()
+
         def pack_to_cpu(tensor):
             if not pin_memory:
                 return (tensor.device, tensor.cpu())
 
-            packed = torch.empty(
-                tensor.size(),
-                dtype=tensor.dtype,
-                layout=tensor.layout,
-                pin_memory=(torch.cuda.is_available() and not tensor.is_sparse))
-            packed.copy_(tensor)
```

Comment:
@albanD Any idea why this approach was chosen for save_on_cpu as opposed to `tensor.to`? Also, is there a reason `non_blocking=True` argument is not passed into `copy_`?

Reply:
As mentioned in the other issue, this split means that we do 2 copies of the Tensor instead of one. So I don't think we should do this.

non_blocking is not useful when moving to the cpu as this operation is always blocking.

Reply:
I see, I wonder why FairScale checkpoint https://github.com/facebookresearch/fairscale/blob/main/fairscale/nn/checkpoint/checkpoint_activations.py#L273 uses it then

Reply:
@albanD I checked some of the copy kernel code: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/Copy.cu#L208-L219, It does look like we still call cudaMemcpyAsync when copying from GPU to CPU when non_blocking=True, so the argument does make some difference? Not sure if I'm missing something here.

Reply:
Well, this is a bug. Nothing on the CPU side ensures proper synchronization.
Running the following on colab does fail:
```python
import torch

N = 100000000

a = torch.ones(N, device="cuda") * 42

for i in range(100):
    a * 2

out = a.to("cpu", non_blocking=True)
t = out.select(0, -1).item()
print(t)


out = a.to("cpu", non_blocking=False)
t = out.select(0, -1).item()
print(t)
```
it prints:
```
0.0
42.0
```

cc @ngimel  this seems to have been here forever. With explicit cases to handle it. Am I missing something here or we should fix this quickly?

Reply:
Note that the doc explicitly mentions that this arg should *only* have an effet when copying CPU -> GPU when the given CPU Tensor is already in pinned memory. And it should have no effet in all other cases.

Reply:
People were (reasonably) asking for a way to do non-blocking copy to cpu, so this was enabled in #46878. And yes, since they explicitly request non-blocking copy, it's their responsibility to ensure proper synchronization. Docs should be updated.